completions/setfacl

```diff
@@ -0,0 +1,92 @@
+# setfacl(1) completion                                    -*- shell-script -*-
+
+_setfacl()
+{
+    local cur prev words cword split
+    # Don't treat user:group as separate words.
+    _init_completion -s -n : || return
+
+    if [[ $cword -ge 2 ]]; then
+        local word
+        for word in "${words[@]:1:$(( cword - 1 ))}"; do
+            if [[ "$word" == '--' ]]; then
+                _filedir
+                return 0
+            fi
+        done
+    fi
+
+    case "$prev" in
+        --restore)
+            _filedir
+            return 0
+            ;;
+
+        -m|--modify|-x|--remove|--set)
+            local other_rules prefix_default mycur
+            other_rules="${cur%"${cur##*,}"}"
+            cur="${cur##*,}"
+            if [[ $cur = d:* || $cur = default:* ]]; then
+                prefix_default=${cur%%"${cur#*:}"}
+                cur="${cur#*:}"
+            else
+                prefix_default=
+            fi
+            mycur="${cur##*:}"
+            if [[ $cur == *:*:*:* ]]; then
+                return
+            elif [[ $cur == *:*:* || $cur == [mo]*:* ]]; then
+                local opts
+                opts='- r w x X rw rx rX wx wX rwx rwX'
+                COMPREPLY=( $( compgen -S ',' -W "$opts" -- "$mycur" ) )
+            elif [[ $cur == u*:* ]]; then
+                local IFS
+                IFS=$'\n'
+                COMPREPLY=( $( compgen -u -S ':' -- "$mycur" ) )
+            elif [[ $cur == g*:* ]]; then
+                local IFS
+                IFS=$'\n'
+                COMPREPLY=( $( compgen -g -S ':' -- "$mycur" ) )
+            else
+                local prefixes
+                prefixes='user: group: mask: other:'
+                if [[ -z $prefix_default ]]; then
+                    prefixes+=' default:'
+                fi
+                COMPREPLY=( $( compgen -W "$prefixes" -- "$mycur" ) )
+            fi
+            if [[ $cur != *:* && -z $prefix_default && "$other_rules" && ${#COMPREPLY[@]} -eq 1 ]]; then
+                COMPREPLY=("${other_rules##*:}${COMPREPLY[0]}")
+            fi
```

Comment:
I haven't checked the actual behavior, but I guess this has the same problem as https://github.com/scop/bash-completion/pull/913#issuecomment-1490140309 when `bind 'set show-all-if-ambiguous on'` is set.

Reply:
This probably wouldn't have been picked up by the `setfacl` team, and you'd know better than I how to fix it.

Reply:
There is no robust way to make the displayed list compact. You'll need to always prefix the `other_rules` in the same way as [PR \#913](https://github.com/scop/bash-completion/pull/913/files#diff-a4757074ff650000804fd3eaabe9b0a9e02e33040ca5b8afd4c0275fc5f3e136R905-R908). For example,

```suggestion
            if [[ $cur != *:* && -z $prefix_default && "$other_rules" ]]; then
                local i
                for i in "${!COMPREPLY[@]}"; do
                    COMPREPLY[i]=${other_rules##*:}${COMPREPLY[i]}
                done
            fi
```

But I haven't tried to understand the part `[[ $cur != *:* && -z $prefix_default && "$other_rules" ]]`, so it might not result in an expected result. Could you explain the condition `[[ $cur != *:* && -z $prefix_default && "$other_rules" ]]`?

Reply:
Re #913:
> I'm not sure why the original only prepended the prefix back when there was only one completion, so I'm worried I'm missing something.

My understanding was that omitting the prefix when there are multiple candidates makes the list more concise, thus easier to visually scan & less likely to cause pagination. It also provides the ability to create richer results, as demonstrated in [this article](https://iridakos.com/programming/2018/03/01/bash-programmable-completion-tutorial#displaying-each-entry-in-a-new-line).

Without your suggestion the behaviour is like this:
```sh
roy@RoyBook:~$ setfacl --set user:w
whoopsie:  www-data:
roy@RoyBook:~$ setfacl --set user:whoopsie:rw
rw,   rwx,  rwX,
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:a
adm:            audio:          avahi:          avahi-autoipd:
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:
-,    r,    rw,   rwx,  rwX,  rx,   rX,   w,    wx,   wX,   x,    X,
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:rw,
default:  group:    mask:     other:    user:
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:rw,other:
-,    r,    rw,   rwx,  rwX,  rx,   rX,   w,    wx,   wX,   x,    X,
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:rw,other:r
```
With it, the previous definition's mode is prepended to the next definition, which is a drop in readability:
```sh
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,
rwX,default:  rwX,group:    rwX,mask:     rwX,other:    rwX,user:
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:a
adm:            audio:          avahi:          avahi-autoipd:
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:
-,    r,    rw,   rwx,  rwX,  rx,   rX,   w,    wx,   wX,   x,    X,
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:adm:rw,
rw,default:  rw,group:    rw,mask:     rw,other:    rw,user:
```

I tried to replicate the prefix loss per the last snippet of [your comment](https://github.com/scop/bash-completion/pull/913#issuecomment-1490140309), but it did not seem to occur. Perhaps this function is not affected due to the slightly different behviour around colons?
```sh
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:avahi
avahi:          avahi-autoipd:
roy@RoyBook:~$ setfacl --set user:whoopsie:rwX,group:av<TAB>
```